src/main/scala/io/opentargets/openfda/stage/OpenFdaEtl.scala

```diff
@@ -15,29 +15,41 @@ object OpenFdaEtl extends LazyLogging {
   def apply(implicit etLSessionContext: ETLSessionContext): DataFrame = {
     implicit val ss: SparkSession = etLSessionContext.sparkSession
 
-    val blackListPath = etLSessionContext.configuration.fda.fdaInputs.blacklist
+    val blackListPath: String = etLSessionContext.configuration.fda.fdaInputs.blacklist
     val chemblPath = etLSessionContext.configuration.fda.fdaInputs.chemblData
     val fdaPath = etLSessionContext.configuration.fda.fdaInputs.fdaData
 
     import ss.implicits._
 
     // load inputs
-    /* load the blacklist terms collect as a list and broadcast the field to
-        all the cluster nodes thus it can be effectively used per row
-     */
-    val bl = broadcast(Loaders.loadBlackList(blackListPath))
-
     // the curated drug list we want
-    val targetList = Loaders.loadTargetListFromChemblDrugs(chemblPath)
-    val drugList = Loaders
-      .loadChemblDrugList(chemblPath)
-      .join(targetList, Seq("chembl_id"), "left")
-      .orderBy(col("drug_name"))
-      .cache()
-
-    // load FDA raw lines
-    val lines = Loaders.loadFDA(fdaPath)
-    val fdasF = lines
+    val drugList: Dataset[Row] = generateDrugList(chemblPath).cache()
+
+    val fdaData = prepareAdverseEventsData(fdaPath)
+
+    // remove blacklisted reactions using a left_anti which is the complement of
+    // left_semi so the ones from the left side which are not part of the equality
+    val fdaDataFilteredByBlackListAndJoinedWithDrug = filterBlacklist(blackListPath, fdaData)
+      .join(drugList, Seq("drug_name"), "inner")
+
+    val groupedByIds = prepareSummaryStatistics(fdaDataFilteredByBlackListAndJoinedWithDrug)
+
+    prepareForMonteCarlo(groupedByIds)
+
+  }
+
+  private def filterBlacklist(blacklistPath: String, df: Dataset[Row])(
+      implicit sparkSession: SparkSession): Dataset[Row] = {
+    val bl = broadcast(Loaders.loadBlackList(blacklistPath))
```

Comment:
We need the whole dataframe in memory in each node in this case don't we since we're effectively using the anti_join as a filter?

Reply:
@JarrodBaker you can even remove the broadcast but when you have on the right a small dataset as we have with blacklisting terms, the best way to operate it (like a lookup table) is _broadcasting_ the full table in each node. this will speed up dramatically the anti join operation as it will reduce to 0 the network shuffling.